pkg/services/sqlstore/user_auth.go

```diff
@@ -101,8 +101,24 @@ func GetUserByAuthInfo(query *models.GetUserByAuthInfoQuery) error {
 		return models.ErrUserNotFound
 	}
 
-	// create authInfo record to link accounts
-	if authQuery.Result == nil && query.AuthModule != "" {
+	// Special case for generic oauth duplicates
+	oauthModule := false
+	if query.AuthModule == "oauth_generic_oauth" && user.Id != 0 {
+		oauthModule = true
+		_, err = x.Id(user.Id).Get(user)
```

Comment:
Think you want to use `GetAuthInfo(authQuery)` and filter by auth module and user id to make sure there exist a row in user_auth for auth_module and user_id.

Reply:
This existing code is a bit confusing. So I looked at how the `GetUserByAuthInfo` is used and it's used here:
https://github.com/grafana/grafana/blob/636be0ed22335c8725412f5832a5bb20fa1c3eab/pkg/services/login/loginservice/loginservice.go#L46-L101

Given that no error is returned and there exist an oauthtoken `updateUserAuth` will be called from there.

So I think that this special case when generic oauth is true we need to `GetAuthInfo` based on auth module and user id which will populate `authQuery.Result`. The `UpdateAuthInfoCommand` shouldn't be needed then. On line 121 if `authQuery.Result` is nil it will insert, otherwise just return the result and eventually the `updateUserAuth` will be called by loginservice.go (if needed)

Let me know if all of this is super confusing :smile:

Reply:
A bit complicated, but feel free to have a look again.